<a href="https://colab.research.google.com/github/sidbhaumik/sidb_datascience_projects.github.io/blob/main/basicNN_with_TF.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Implementing a Neural network with TensorFlow

#Step 1: Set up the environment

In [ ]:
pip install tensorflow

Once TensorFlow is installed, import the necessary libraries to begin building the neural network.

In [ ]:
import tensorflow as tf
from tensorflow.keras import layers, models

#Step 2: Load and preprocess the dataset:
For this activity, We will use the Fashion MNIST dataset, which consists of 28x28 (70,000) grayscale images of fashion items, with 10 different classes (categories). TensorFlow provides a built-in utility to load this dataset.

In [ ]:
# Load the Fashion MNIST dataset
(train_images, train_activityels), (test_images, test_activityels) = tf.keras.datasets.fashion_mnist.load_data()

# Normalize the pixel values to be between 0 and 1
train_images = train_images / 255.0
test_images = test_images / 255.0

29515/29515 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
26421880/26421880 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
5148/5148 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
4422102/4422102 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


##Explanation:
Normalizing pixel values to a range of 0 to 1 (instead of from 0 to 255) ensures efficient model training and helps avoid issues with large input values that can negatively impact the optimization process.

#Step 3: Define the neural network model
Now you will define the architecture of the neural network using the TensorFlow Keras API. The network will consist of:

1. An input layer that flattens the 28 × 28 image into a one-dimensional vector.

2. A hidden layer with 128 neurons and the ReLU activation function.

3. An output layer with 10 neurons (one for each fashion class) using softmax activation.

In [ ]:
# Define the model
model = models.Sequential([
    layers.Flatten(input_shape=(28, 28)),  # Input layer to flatten the 2D images
    layers.Dense(128, activation='relu'),  # Hidden layer with 128 neurons
    layers.Dense(10, activation='softmax') # Output layer with 10 classes
])

/usr/local/lib/python3.11/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


## Explanation:

1. Flatten layer: Converts the 28 × 28 matrix into a one-dimensional array of 784 features.

2. Dense layers: A fully connected layer with 128 neurons and a ReLU activation function for the hidden layer, followed by an output layer with 10 neurons (one per class) that uses softmax to output probabilities.

# Step 4: Compile the model:

After defining the architecture, you need to compile the model. During compilation, you specify:

 1. The optimizer: For this activity, we will use Adam, a widely used optimizer that adjusts learning rates during training.

 2. The loss function: Since this is a classification task, we will use sparse categorical crossentropy.

 3. Metrics: We will track accuracy to monitor model performance.

In [ ]:
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

## Explanation
1. Adam optimizer: Adam is a popular optimizer for deep learning models due to its adaptive learning rate, which often leads to faster convergence.

2. Sparse categorical cross-entropy: This is a suitable loss function when dealing with integer labels (i.e., when each label is a class index).

3. Accuracy metric: The accuracy metric helps you monitor how well the model is predicting the correct classes during training and evaluation.

# Step 5: Train the model
With the model compiled, you can now train it on the Fashion MNIST dataset. We will train the model for 10 epochs with a batch size of 32.

In [ ]:
# Train the model
model.fit(train_images, train_activityels, epochs=10, batch_size=32)

Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 3ms/step - accuracy: 0.7805 - loss: 0.6417
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8604 - loss: 0.3926
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8783 - loss: 0.3385
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.8848 - loss: 0.3135
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8898 - loss: 0.2969
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.8956 - loss: 0.2801
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8963 - loss: 0.2751
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9055 - loss: 0.2545
Epoch 9/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 3ms/step - accuracy: 0.9097 - loss: 0.2432
Epoch 10/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9133 - loss: 0.2308


## Explanation
1. Epochs: This refers to the number of times the model will go through the entire training dataset. Ten epochs is a good starting point for this task.

2. Batch size: This refers to the number of samples processed before the model’s weights are updated. A batch size of 32 is a common choice.

# Step 6: Evaluate the Model
Once the model is trained, you can evaluate its performance on the test data. This will give you a sense of how well the model generalizes to new, unseen data.

In [ ]:
# Evaluate the model on the test set
test_loss, test_acc = model.evaluate(test_images, test_activityels)

print(f'Test accuracy: {test_acc}')

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8842 - loss: 0.3278
Test accuracy: 0.8849999904632568


## Explanation
The test accuracy metric provides insight into how well the model performs on the test dataset. You should aim for an accuracy of around 85–90 percent for this particular dataset.

Test loss and accuracy: The test loss indicates how well the model's predictions match the true labels on the test data, while the test accuracy shows the percentage of correct predictions.

# Step 7: Experimentation
After successfully implementing the basic neural network, we can experiment with the model. Here are a few ideas:

1. Add more hidden layers to make the network deeper.

2. Change the number of neurons in the hidden layer.

3. Try different activation functions, such as tanh or sigmoid, and observe their impact on the model’s performance.

4. Adjust the optimizer: Test how using SGD instead of Adam affects training and accuracy.

In [ ]:
model = models.Sequential([
    layers.Flatten(input_shape=(28, 28)),
    layers.Dense(128, activation='relu'),
    layers.Dense(64, activation='relu'),  # Additional hidden layer with 64 neurons
    layers.Dense(10, activation='softmax')
])

In [ ]:
# Compile the model
model.compile(optimizer='sgd',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [ ]:
# Train the model
model.fit(train_images, train_activityels, epochs=10, batch_size=32)

Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.6679 - loss: 1.0456
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.8202 - loss: 0.5143
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8383 - loss: 0.4601
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8493 - loss: 0.4330
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8561 - loss: 0.4123
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 2ms/step - accuracy: 0.8631 - loss: 0.3951
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8637 - loss: 0.3837
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.8679 - loss: 0.3720
Epoch 9/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8719 - loss: 0.3620
Epoch 10/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8766 - loss: 0.3538


In [ ]:
# Evaluate the model on the test set
test_loss, test_acc = model.evaluate(test_images, test_activityels)

print(f'Test accuracy: {test_acc}')

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8648 - loss: 0.3847
Test accuracy: 0.8623999953269958


With SGD optimizer and additional neurons, the accuracy dropped to .86 and the loss increased to .38.